src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,638 @@
+        def convert_element(value):
+            name = 'draw(%selements)' % (prefix,)
+            try:
+                return np.array([value], dtype=dtype)[0]
+            except TypeError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r of type %s to dtype %s' % (
+                        name, value, type(value).__name__, dtype.str
+                    )
+                )
+            except ValueError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r to type %s' % (
+                        name, value, dtype.str,
+                    )
+                )
+        elements = elements.map(convert_element)
+    assert elements is not None
+
+    return elements, dtype
+
+
+class ValueIndexStrategy(st.SearchStrategy):
+    def __init__(self, elements, dtype, min_size, max_size, unique):
+        super(ValueIndexStrategy, self).__init__()
+        self.elements = elements
+        self.dtype = dtype
+        self.min_size = min_size
+        self.max_size = max_size
+        self.unique = unique
+
+    def do_draw(self, data):
+        result = []
+        seen = set()
+
+        iterator = cu.many(
+            data, min_size=self.min_size, max_size=self.max_size,
+            average_size=(self.min_size + self.max_size) / 2
+        )
+
+        while iterator.more():
+            elt = data.draw(self.elements)
+
+            if self.unique:
+                if elt in seen:
+                    iterator.reject()
+                    continue
+                seen.add(elt)
+            result.append(elt)
+
+        dtype = infer_dtype_if_necessary(
+            dtype=self.dtype, values=result, elements=self.elements,
+            draw=data.draw
+        )
+        return pandas.Index(result, dtype=dtype, tupleize_cols=False)
+
+
+DEFAULT_MAX_SIZE = 10
+
+
+@st.cacheable
+@st.defines_strategy
+def range_indexes(min_size=0, max_size=None):
+    """Provides a strategy which generates an :class:`~pandas.Index` whose
+    values are 0, 1, ..., n for some n.
+
+    Arguments:
+
+    * min_size is the smallest number of elements the index can have.
+    * max_size is the largest number of elements the index can have. If None
+      it will default to some suitable value based on min_size.
+
+    """
+    st.check_valid_interval(min_size, max_size, 'min_size', 'max_size')
+    if max_size is None:
+        max_size = min_size + DEFAULT_MAX_SIZE
+    return st.integers(min_size, max_size).map(pandas.RangeIndex)
+
+
+@st.cacheable
+@st.defines_strategy
+def indexes(
+    elements=None, dtype=None, min_size=0, max_size=None, unique=True,
+):
+    """Provides a strategy for producing a :class:`pandas.Index`.
+
+    Arguments:
+
+    * elements is a strategy which will be used to generate the individual
+      values of the index. If None, it will be inferred from the dtype. Note:
+      even if the elements strategy produces tuples, the generated value
+      will not be a MultiIndex.
```

Comment:
just to clarify, it'll be an index where each index value is a tuple, yeah?

Reply:
Yup. I'll update to clarify.